source/extensions/filters/listener/tls_inspector/tls_inspector.cc

```diff
@@ -354,6 +606,48 @@ void Filter::createJA3Hash(const SSL_CLIENT_HELLO* ssl_client_hello) {
   }
 }
 
+void Filter::createJA4Hash(const SSL_CLIENT_HELLO* ssl_client_hello) {
+  if (!config_->enableJA4Fingerprinting()) {
+    return;
+  }
+
+  std::string fingerprint;
+
+  // Protocol type (t for TLS, q for QUIC, d for `DTLS`)
+  // In this implementation we only handle TLS
+  absl::StrAppend(&fingerprint, "t");
```

Comment:
I think it is more efficient if you do a single `absl::StrAppend` with all of the below values, because it doesn't have to resize the `std::string` destination multiple times. You can still have one arg per line with the comments above each one, as you do now.